[SymmetricMemory] add support for cuStreamWriteValue32

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136488

cuStreamWriteValue efficiently combines the issuing of a system-level fence with the update of a single memory location. It is highly suitable for inter-stream progress sharing (e.g., all_gather_with_progress).

Exposing it via SymmetricMemory allows users to more easily implement efficient progress-aware matmuls in triton ([xformers example](https://github.com/facebookresearch/xformers/blob/main/xformers/ops/_triton/sequence_parallel_fused_kernels.py)).

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Commit:
[SymmetricMemory] add support for cuStreamWriteValue32

cuStreamWriteValue efficiently combines the issuing of a system-level
fence with the update of a single memory location. It is highly suitable
for inter-stream progress sharing (e.g., all_gather_with_progress).

[ghstack-poisoned]

eqy commented on test/distributed/test_symmetric_memory.py:
```diff
@@ -502,6 +502,29 @@ def test_multimem_one_shot_all_reduce(
         self.assertTrue(res.eq(self.world_size).all().item())
         dist.destroy_process_group()
 
+    @skipIfRocm
```
Does this need any additional decorators for e.g., min number of GPUs or P2P support?

yifuwang replied:
Good catch. The test requires >=2 GPUs to run.

eqy commented on test/distributed/test_symmetric_memory.py:
```diff
@@ -502,6 +502,29 @@ def test_multimem_one_shot_all_reduce(
+            symm_mem.stream_write_value32(
+                int(tensor.data_ptr()) + i * tensor.element_size(), 1
+            )
+            self.assertTrue(torch.allclose(tensor, expect[i]))
```
Consider using `self.assertEqual` or `torch.testing.assert_close` for debuggable error messages

Commit:
Update on "[SymmetricMemory] add support for cuStreamWriteValue32"


cuStreamWriteValue efficiently combines the issuing of a system-level fence with the update of a single memory location. It is highly suitable for inter-stream progress sharing (e.g., all_gather_with_progress).

Exposing it via SymmetricMemory allows users to more easily implement efficient progress-aware matmuls in triton ([xformers example](https://github.com/facebookresearch/xformers/blob/main/xformers/ops/_triton/sequence_parallel_fused_kernels.py)).

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

Commit:
Update on "[SymmetricMemory] add support for cuStreamWriteValue32"


cuStreamWriteValue efficiently combines the issuing of a system-level fence with the update of a single memory location. It is highly suitable for inter-stream progress sharing (e.g., all_gather_with_progress).

Exposing it via SymmetricMemory allows users to more easily implement efficient progress-aware matmuls in triton ([xformers example](https://github.com/facebookresearch/xformers/blob/main/xformers/ops/_triton/sequence_parallel_fused_kernels.py)).

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge
